### ISSUE TITLE
Add LightGBM learners to MMLSpark

### ISSUE THREAD

**Issue by imatiach-msft** [Azure/mmlspark#173]
Add LightGBM learners to MMLSpark as spark Estimators and Transformers

**Comment by David082** [comment]
Where can I found the package com.microsoft.ml.lightgbm？

**Comment by imatiach-msft** [comment]
Hi @David082, 
We will be uploading it to maven central as soon as we are done.  It is currently in a private repository.
Thank you, Ilya

**Comment by imatiach-msft** [comment]
@David082 I have checked in the code to generate the SWIG wrappers into the LightGBM repo.
To build, you just need to run:

mkdir build ; cd build
cmake -DUSE_SWIG=ON ..
make -j4

This will generate a jar file containing the LightGBM c_api wrapped by SWIG.
You will need to load the native .so files as in the lightgbm branch before calling into the Java API.
I am currently working on the LightGBM estimator in the lightGBM branch in mmlspark - I am currently doing performance testing and optimization.

**Comment by imatiach-msft** [comment]
@David082 LightGBM is now available in maven central:
https://repo.maven.apache.org/maven2/com/microsoft/ml/lightgbm/lightgbmlib/
You can import it with sbt via:
"com.microsoft.ml.lightgbm" %  "lightgbmlib" % "2.0.120"

**Comment by David082** [comment]
Thank you!

**Comment by ezerhoun** [comment]
Hello,

I saw that `LightGBMClassifier` has been added [LightGBMClassifier](https://mmlspark.azureedge.net/docs/0.10.dev26+4.g8fa9cc3a4/pyspark/LightGBMClassifier.html) Do you plan to add also a LightGBMRegressor ?

Thank you very much

**Comment by imatiach-msft** [comment]
Yes, I will add a separate LightGBMRegressor in the near future, I think it should be trivial.  I'm also going to extend the current classifier to support multiclass in addition to binary labels, this should be easy to do as well.  Adding the ranker may take some planning and thought though in order to make it fit into the spark ecosystem.

**Comment by ezerhoun** [comment]
@imatiach-msft Thank you !

**Comment by troszok** [comment]
@imatiach-msft great job with the classifier, thank you very much! We are waiting now for the regressor as well.

**Comment by imatiach-msft** [comment]
@troszok Thanks!  The classifier has been checked in and I have created a PR to add the regressor here:
https://github.com/Azure/mmlspark/pull/249

**Comment by imatiach-msft** [comment]
@troszok the LightGBM regressor has been merged into master, we would love to hear any feedback you might have for the interface and the parameters we currently expose - in my testing the learner was 10-30% faster in execution time than spark's GBTClassifier and had 15% better AUC on the Higgs dataset

**Comment by ezerhoun** [comment]
@imatiach-msft Thank you for adding LightGBM regressor !

I am trying to train a LightGBM Quantile regressor. However, I don't see how to pass the quantile parameter (`alpha` parameter in LightGBM). Do you have an example of how to do it ?

**Comment by imatiach-msft** [comment]
@ezerhoun thank you for the great feedback!  I've created a PR which adds the alpha parameter for lightGBMRegressor and a pyspark notebook example for how to use lightGBM with quantile regression:
https://github.com/Azure/mmlspark/pull/254
It's currently in code review.
If you need to add more params but can't wait for us to add them to the learner you can also put the param=value inside the "application" parameter, eg:

df = ...
reg = LightGBMRegressor(application='quantile alpha=0.1')
model = reg.fit(df)

it must be in the format that lightgbm accepts

**Comment by ezerhoun** [comment]
@imatiach-msft Thank you for your quick reply and PR !
Moreover, I have couple of other questions:
- I would like to use the `jar` produced by a PR, but after the PR being merged, the `jar` is deleted. Do you have an existing jar in https://mmlspark.azureedge.net/maven which contains the lastest code (`SNAPSHOT`) ?
- Will it be possible to save a LightGBM model and load it using the python or java implementation (via SWIG) ?

**Comment by imatiach-msft** [comment]
@ezerhoun Is the scenario to train in spark but score in a java environment?  In that case, I think it is possible
1.) For now, you can use the maven coordinates that are produced through the build but you will need to add our resolver or repository (https://mmlspark.azureedge.net/maven) -- as soon as we have a new release you will be able to use coordinates from maven central as explained in the main page
2.) Yes, it should be possible - when you save the model in spark in one of the folders there will be a text file representing the model which you can then parse using either LightGBM's native C++ api, LightGBM's load model in python or R, or the load model in the Java jar that is in maven

**Comment by imatiach-msft** [comment]
@ezerhoun for now, you can use the following options when running spark:
--packages com.microsoft.ml.spark:mmlspark_2.11:0.11.dev31+1.gf23ab7852
--repositories https://mmlspark.azureedge.net/maven
please let me know if you have questions or comments

**Comment by ezerhoun** [comment]
@imatiach-msft Thank you for your replies . Indeed the idea would be to train in Spark and test in Java.
I don't know if it is the right place to ask but I have the following error : 

```
Py4JJavaError: An error occurred while calling o673.fit.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 4 in stage 47.1 failed 4 times, most recent failure: Lost task 4.3 in stage 47.1 (TID 3569, 10.233.104.114, executor 33): java.lang.Exception: Network init call failed in LightGBM with error: Machine list file doesn't contain the local machine
```
Should I do as in [LightGBM Parallel Learning Guide](https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html) and create a `mlist.txt`

Thanks !

**Comment by imatiach-msft** [comment]
@ezerhoun this might be a bug -- do you see this line printed out:
log.info(s"Nodes used for LightGBM: $nodes")
What do you see as the nodes?  It should contain a list of hostname:port which is used to initialize lightgbm when calling NetworkInit.  Are you running this on a spark cluster or a VM?  What is the spark cluster setup?  Currently I am retrieving the nodes in getExecutors(...) method via BlockManager:
``` Scala
val blockManager = BlockManagerUtils.getBlockManager(dataset)
    blockManager.master.getMemoryStatus.flatMap({ case (blockManagerId, _) =>
      if (blockManagerId.executorId == "driver") None
      else Some((blockManagerId.host, blockManagerId.executorId.toInt))
    }).toArray
```

**Comment by imatiach-msft** [comment]
@ezerhoun just to note in case it wasn't clear, the error you are seeing, specifically:
"Network init call failed in LightGBM with error: Machine list file doesn't contain the local machine"
Is coming from this call in the scala code:
LightGBMUtils.validate(lightgbmlib.LGBM_NetworkInit(nodes, localListenPort,
        LightGBMConstants.defaultListenTimeout, numNodes), "Network init")
In the file TrainUtils.scala in method trainLightGBM.
I think the problem might be with the "nodes" parameter that is passed in.  For local vs cluster the logic is slightly different, if you are using a different compute environment than what I have tested it's possible that the nodes may need to be fetched in a different way, so I may need more information about your compute context.

**Comment by imatiach-msft** [comment]
The logs should look something like this (taken from my logs on a spark HDI yarn cluster):

18/03/05 17:32:08 INFO LightGBMClassifier: Nodes used for LightGBM: 10.0.0.10:12401,10.0.0.11:12410,10.0.0.13:12403,10.0.0.14:12406,10.0.0.15:12407,10.0.0.16:12404,10.0.0.17:12408,10.0.0.18:12405
--

**Comment by imatiach-msft** [comment]
@ezerhoun I believe I have figured out the issue, there was a bug in my new code.  I've created a new build, will send you the new package when it is finished.

**Comment by imatiach-msft** [comment]
@ezerhoun the fixed package is:
--packages com.microsoft.ml.spark:mmlspark_2.11:0.11.dev31+1.g4ad3176c9
--repositories https://mmlspark.azureedge.net/maven.

**Comment by troszok** [comment]
@imatiach-msft  great news!!! I forgot to check the progress here. We will try to do POC with new model and we will let you know!

**Comment by ezerhoun** [comment]
@imatiach-msft I tried the new package and I still have the same problem.

**Comment by imatiach-msft** [comment]
@ezerhoun could you give me information about your cluster setup?  I have validated the following scenarios:
1.) Local tests (running under local[*])
2.) Running on HDI with yarn and # partitions < # executors (this was failing before the latest PR)
3.) Running on HDI with yarn and # partitions >= # executors (this worked before but my PR broke it yesterday, and I fixed it with the latest package I sent above)
I am wondering if there is some cluster setup that does not work with the current code.

**Comment by imatiach-msft** [comment]
@ezerhoun it might be faster to discuss over skype.  You can email:
mmlspark-support@microsoft.com
and I can send you a link to a skype meeting

**Comment by mhamilton723** [comment]
@imatiach-msft should this be closed?

**Comment by iainmillar23** [comment]
Any idea when the mmlspark v0.12 that includes the new LightGBMClassifier will be released to maven central?

**Comment by elibarzilay** [comment]
@iainmillar23 Hopefully by Monday (ish).

**Comment by imatiach-msft** [comment]
@mhamilton723 I think we can leave this open for now, maybe for at least 2 release cycles, since I would still like to add multiclass classification and possibly the ranker

**Comment by ekaterina-sereda-rf** [comment]
Hi!

> --packages com.microsoft.ml.spark:mmlspark_2.11:0.11.dev31+1.g4ad3176c9
> --repositories https://mmlspark.azureedge.net/maven.

I tried to use this packages and was not successful on downloading dependency. Did something change?

**Comment by elibarzilay** [comment]
We're going to release 0.12 very soon, it'll have that with more thing.
(It got delayed with some technical issues.)

**Comment by ekaterina-sereda-rf** [comment]
Thank you, waiting for that!

**Comment by ekaterina-sereda-rf** [comment]
Hi! I have one more question. I found development version that you posted earlier, but now i faced problem, that i can't pass all parameters that i need.
`max_bin, bagging_fraction,bagging_freq,bagging_seed,feature_fraction,max_depth,min_sum_hessian_in_leaf`
i tried to change your code, and added them to parent parameter list to the `trait LightGBMParams`. so i see parameters set now, but i don't see effect on training. Are you going to support them? do you have quick advice for me what also could be changed in lib so these parameters will be taken in account?

**Comment by Ragavenderan** [comment]
Can we train this in Azure batch?

**Comment by imatiach-msft** [comment]
@peay this should supposedly not be an issue if you are using the LightGBMClassifier or LightGBMRegressor -- from here we get all "executor:ports" and pass that to the network init method in LightGBM:

https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMUtils.scala#L57

so supposedly if you have two executors on the same VM you should be using the same executor ip address twice and two different port numbers (the port value is 12400 + executor id).  If you are seeing issues then that must mean there is a bug in this logic.  Is there a way I could get a repro of your setup?  I noticed another user above encountered a similar issue but unfortunately was unable to give me a repro.

**Comment by imatiach-msft** [comment]
"I gave the bindings a try"

@peay sorry, maybe I misunderstood, are you specifically using the scala learners or the Java bindings from the LightGBM maven artifact directly?
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMClassifier.scala
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMRegressor.scala

**Comment by ekaterina-sereda-rf** [comment]
@imatiach-msft sorry for long absence. So here is my changes 
https://gist.github.com/ekaterina-sereda-rf/b7b718da8a14f755b2dfcea62cf54123
Question is - is it enough to pass it engine? (of course I set them in my code after it) 
And also I have question about performance. Do you have some statistic about how long should model should train on different amount of data, with different params so on? 
p.s. i wanted to do pull request, but i saw that i should subscribe something, so on - so i though this way is more easy. but if you'll help me to start i can do it with pleasure.
Thank you!

**Comment by imatiach-msft** [comment]
@ekaterina-sereda-rf this looks great!  Note you will need to pass those params to the underlying LightGBM C++ api here (similar to the other parameters):
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/TrainParams.scala#L15
If you like I can create a pull request with your changes, but if you are interested in doing this you can also contribute, you just need to fork the mmlspark repo, make the code changes and then send out a PR.  Please let me know what you prefer.  Thanks!

**Comment by imatiach-msft** [comment]
@ekaterina-sereda-rf one nit-pick, could you remove the "_" in the parameter names to be consistent with the other parameters?
Also, for your question about performance, I've only tested this on the Higgs dataset as was recommended by @guolinke, and the results showed that LightGBM was faster than GBTClassifier by 10-30% and had a better AUC by 15%.  I'm sure we will need to do more performance testing in the future, but if you see any issues on any datasets please let me know.

**Comment by katerinagl** [comment]
Thank you for your answers! yes, I did not add it to trainParams before, so I think it will help. I created PR https://github.com/Azure/mmlspark/pull/282 , please have a look and let me know if I need to fix something.
And about performance I'll let you know when will have fresh results.

**Comment by peay** [comment]
Ah, I see -- I was a bit too quick to assume executor collisions then: the fact that executors were running on the same node may have been a coincidence.

I've investigated a bit more, and here is a more accurate description of what I run into:

- I load and persist some data with Spark
- Call `mmlspark.LightGBMRegressor` (from Python)
- A job is scheduled, with a couple dozens of `persist` stages, and a final `reduce` stage
- Then, what happens appears to be a bit random. Sometimes, half of the `persist` stages will succeed and the next one will fail. Sometimes, they will all succeed and the final `reduce` will fail.

In all cases, the error is

```
Py4JJavaError: An error occurred while calling o5186.fit.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 8 in stage 213.1 failed 4 times, most recent failure: Lost task 8.3 in stage 213.1 (TID 11285, 10.233.78.219, executor 0): java.lang.Exception: Network init call failed in LightGBM with error: Binding port 12400 failed
	at com.microsoft.ml.spark.LightGBMUtils$.validate(LightGBMUtils.scala:16)
	at com.microsoft.ml.spark.TrainUtils$.trainLightGBM(TrainUtils.scala:141)
	at com.microsoft.ml.spark.LightGBMRegressor$$anonfun$1.apply(LightGBMRegressor.scala:70)
	at com.microsoft.ml.spark.LightGBMRegressor$$anonfun$1.apply(LightGBMRegressor.scala:70)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$6.apply(objects.scala:186)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$6.apply(objects.scala:183)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:827)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:827)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:99)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:322)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

The port varies -- executor 0 here, but sometimes others. Running about 10 executors most of the time. Pinning it down to one executor only, I am not able to reproduce but it may also be by chance.

I will try to investigate more this week, and get to the logs of the executors.

This is a Spark standalone cluster, running Spark 2.1.1 on-premise on top of kubernetes.

**Comment by imatiach-msft** [comment]
@peay one possibility is that port number, 12400, is actually in use - currently there is no logic to find open ports on the workers and each worker just opens port ("defaultListenerPort" + executor_id), this parameter is defined here:
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMParams.scala#L19
And the logic to calculate the port number is here:
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMUtils.scala#L91
One workaround may be to set the default listener port parameter to a slightly higher value, eg 12420, and see if that makes any difference.
Also, the way I get the executors is via the blockManager if the number of executors <= num partitions after doing a coalesce on the dataset to the number of executors:
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMUtils.scala#L59
Then this is followed by a map partitions to do the actual training.
Otherwise, if num executors > num partitions, I do a map partitions twice, the first time to get the executor ids and the second time to do the training logic:
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMUtils.scala#L128
This would only work if the same executors are run twice, which in my testing was always the case.  How many partitions do you have in your dataset?  If it is >= # executors, could you also try to repartition to # executors - 1 partitions and see if the training works?  I think it may be more reliable to do the map partitions twice but there is an execution time performance cost unfortunately.

**Comment by imatiach-msft** [comment]
@iainmillar23 @ekaterina-sereda-rf v0.12 has been released for spark 2.2.1 if you want to try it out

**Comment by imatiach-msft** [comment]
@peay I'm going to try to reproduce it today, would you be able to give me your yml file too?  I will try to create the cluster on AKS (I've been using HDInsight so far for testing).

**Comment by iainmillar23** [comment]
Does v0.12 required Spark 2.2.1 or will it work with Spark 2.1.0?

**Comment by imatiach-msft** [comment]
@iainmillar23 It should work with 2.1.0, although our source code is using 2.2.1 they are backwards compatible.
However once we move to 2.3.0 it won't be because OneHotEncoder was changed to OneHotEncoderEstimator.

**Comment by imatiach-msft** [comment]
@iainmillar23 @ekaterina-sereda-rf sorry, it looks like we haven't yet uploaded to spark packages, we will be doing that later today - so some of the installation instructions might not work yet.

**Comment by imatiach-msft** [comment]
@peay I tried the latest 0.12 on HDInsight with LightGBMClassifier on Higgs dataset and it worked fine, I'm still trying to create the AKS cluster.  One useful log to look for in the yarnui might be the list of nodes and ports:

18/04/17 16:03:48 INFO LightGBMClassifier: Nodes used for LightGBM: 10.0.0.10:12454,10.0.0.11:12458,10.0.0.13:12453,10.0.0.14:12451,10.0.0.14:12459,10.0.0.15:12455,10.0.0.16:12452,10.0.0.16:12460,10.0.0.17:12457,10.0.0.18:12456
--

This might help diagnose the issue.  You should see 10 entries if you are using 10 executors.

**Comment by imatiach-msft** [comment]
@iainmillar23 @ekaterina-sereda-rf the spark package has been uploaded, so the release is complete, in case you want to try out LightGBMClassifier/Regressor

**Comment by ekaterina-sereda-rf** [comment]
@imatiach-msft Thank you very much :) already tried it today. and have new questions.
1. in general, seems we also saw  `Network init call failed in LightGBM with error: Binding port 12400 failed
	at com.microsoft.ml.spark.LightGBMUtils$.validate(LightGBMUtils.scala:16) ` in some cases, when we used more then one machine - but I can't describe conditions now.(will do if i see it again)
2. [Here](https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMUtils.scala#L61)  and [here](https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/main/scala/LightGBMRegressor.scala#L57)  - seems you paralelize everything by number of executors but not by number of cores - does it have some explanation in field of inner calculation? because i changed it manually to number of cores we use - and it works much much faster.  now we testing accuracy. Any way, thank you very much for your great job.

**Comment by ekaterina-sereda-rf** [comment]
And one more thing. Do you have some insights about memory usage? for example i have 10gb of data for training and 100 features - how much memory should i have to finish training?(i don't know if i formulated correctly, i understand that there is a lot of factors, but if you have some examples that you can share - it could really help me)

**Comment by imatiach-msft** [comment]
@ekaterina-sereda-rf yes memory usage is definitely an issue, as the data needs to be copied to lightgbm into native code in order to train the distributed learner - so basically we create a second copy of the dataset on each partition.  I assume the 10 GB dataset would use about 10 GB memory for spark and another 10 GB memory for lightGBM, so approximately 20 GB of data.  Are you seeing any memory issues outside of 2X cost?

**Comment by imatiach-msft** [comment]
@ekaterina-sereda-rf yes, currently I am parallelizing by number of executors.  Performance may be improved more by parallelizing by number of cores.  In my testing I used a cluster with one core per executor.  I discussed with @mhamilton723 and it sounds like one executor can process multiple partitions at once on all threads.  If that is the case then when doing network init we can pass each thread instead of each executor to get a significant speedup in cases where there are multiple cores per executor.  I am also considering modifying the code to either push rows in parallel into lightgbm, but this may involve some changes in the lightgbm code, or have singletons per executor and push native data there on all threads in one map partitions and then do a separate map partitions to push that native data to create a lightgbm dataset - this might be more robust but I'm not sure if the performance would be better as we would need to do two map partitions calls, and the code would be much more complex.

**Comment by katerinagl** [comment]
Hi :) so, i run training, on pure your code (without my changes). i have no data prep - i have ready data with features and load it from parquet, it's only step on my side. i have 1 executor with 400gb of ram and 64 core, and for last 5+ hours i see this line in logs. `INFO LightGBMRegressor: LightGBM worker listening on: 12450` - so as i see iterations even did not start. for sure as we discuss above only one core occupied, and there is more then 200 gb of RAM occupied (to question about memory), my data set is about 20gb. maybe i do something wrong? maybe there is a way to see more logs? something on c++ side?

**Comment by katerinagl** [comment]
@iainmillar23 i use this resolver - and v0.12 was loaded `resolvers += "MMLSpark Repo" at "https://mmlspark.azureedge.net/maven"`

**Comment by imatiach-msft** [comment]
@katerinagl @iainmillar23 it should eventually be in maven central like the other versions here:
https://mvnrepository.com/artifact/Azure/mmlspark
@elibarzilay I believe all spark packages are added to maven central, is this correct?  It may take an extra day or so for it to show up as we just released a day ago.

**Comment by imatiach-msft** [comment]
@katerinagl interesting, I'm guessing there must be a bug.   I would need to reproduce the issue somehow to understand what is going on.  If you use a smaller dataset do you see the same behavior?  Our sample notebooks in the docker run on a single core so LGBM would fail in those, and I do have some tests (the fuzzing tests) that run on a single executor:
https://github.com/Azure/mmlspark/blob/master/src/lightgbm/src/test/scala/VerifyLightGBMClassifier.scala#L111 
If you run with Spark ML's GBTClassifier does the code work?  Does using 64 executors, one per core, mitigate the issue?

**Comment by imatiach-msft** [comment]
@katerinagl @iainmillar23 the package is in maven central now - you can see v0.12 there now:
https://mvnrepository.com/artifact/Azure/mmlspark
It seems it just takes a couple of days.

**Comment by imatiach-msft** [comment]
@katerinagl @peay @ezerhoun  I tried running lightGBM in an azure databricks cluster with a configuration of 3 worker nodes, 3 executors and 8 cores per executor.  I was hoping to reproduce the problem with network init by running multiple cores per executor (as opposed to HDI case where there was 1 core per executor) but LightGBM completed.  However, I found that it was 2X slower than GBTClassifier in this case (160 seconds GBT vs 311 seconds LGBM), I think it only uses one of the cores.  I think the issue @katerinagl ran into when running just one executor is related, I would recommend using as many executors as cores if you use LightGBM.  I think I might be able to figure out a way to optimize and fully utilize all cores per executor when there is more than 1 by spawning socket connections on each thread and opening up multiple ports per executor.

**Comment by katerinagl** [comment]
Hi:) so when i run multiple executors on one machine i run into the problem with ports, so i need to change this moment, and bind as many addresses as many executors i have. this works. but when i split one machine into, several executors, i fell in problems with memory. 
so for me is still big question what is the difference from algorithm/calculation perspective when i run several executors or i run several cores on one executor - i was succeeded with both variants after my changes, but in first case i had bad accuracy, on small dataset i had all predictions 0.0 (and the same case, same data with 1 core made at least some predictions) and with second case(several executors) i did not complete testing yet. i used 4 executors and it made some results. i don't know what about 4 cores. i hope there is not too much mess here. in gerenal if you can provide me datasets that you use for testing.(i tried to find it in repo, but did not see it)

**Comment by imatiach-msft** [comment]
@katerinagl I've put the azure databricks and HDInsight notebooks here for the Higgs dataset in my branch (note I will not be checking these in):
https://github.com/imatiach-msft/mmlspark/tree/ilmat/higgs-notebooks/notebooks/higgs
You can get the Higgs csv dataset here:
https://github.com/guolinke/boosting_tree_benchmarks/tree/master/data
Specifically this link:
https://archive.ics.uci.edu/ml/datasets/HIGGS

My TODOs right now are:
1.) I think I need to spend time investigating how I can optimize for the multi-core per executor scenario, my intuition is that I need to open a socket per thread instead of per executor as I am doing right now when calling network init; I'm not sure if each call inside the mapPartitions is per executor or per thread - it sounds like it is per thread.  Also, another thing I might try out is keeping a global singleton per executor where I would push the data to (and I would not need to coalesce the data in that case), and then doing the training afterwards
2.) I also still need to find a way to repro the network_init failures some people are seeing.  The only way that I know to repro it is to run multiple lightGBM learners in succession -- the ports take some time to free, but there is an easy mitigation which is to give different base port values in the params when sweeping over the learner.  Eventually it would be good to implement a way to search for free ports on the executors to reduce the chance of failure during network init.

**Comment by elibarzilay** [comment]
Yes, I think that spark packages will put the jar file into maven central, but probably not the lightgbm thing?

**Comment by imatiach-msft** [comment]
@katerinagl the length of this thread is quite long, and I'm now noticing some performance issues on github.  I've created a separate thread to track the performance issues with lightGBM when using clusters configured with multiple cores per executor here:
https://github.com/Azure/mmlspark/issues/292
I would also like to eventually get a repro of the network init failures some people are having, but I haven't been able to reproduce this yet.
